debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsStreamProducerIT.java

```diff
@@ -1016,8 +1016,9 @@ public void shouldReceiveHeartbeatAlsoWhenChangingNonWhitelistedTable() throws E
                 "INSERT INTO s1.b (bb) VALUES (22);";
 
         // streaming from database is non-blocking so we should receive many heartbeats
-        final int expectedAtMostStartHeartbeats = 10;
-        final int expectedHeartbeats = 5;
+        // plus keepalive messages also generate heartbeats
+        final int expectedAtMostStartHeartbeats = 500;
```

Comment:
Why is it that many heartbeats are expected here now?

Reply:
I'll try to figure it out

Reply:
@gunnarmorling In current implementation heartbeat is sent when there was a change in any table and `RecordsStreamProducerIT#shouldReceiveHeartbeatAlsoWhenChangingNonWhitelistedTable` should expect only one heartbeat (produced by `INSERT INTO s1.a (aa) VALUES (11);`) before change record (produced by `INSERT INTO s1.b (bb) VALUES (22);`)

Why `expectedAtMostStartHeartbeats = 10` was used? `expectedAtMostStartHeartbeats = 1` should work fine. I have tested it with decoderbufs, may try it later with other plugins. Maybe one of the plugins produces some additional messages.

Anyway when change record is produced a lot of heartbeats produced after that. The first reason is because of `.with(Heartbeat.HEARTBEAT_INTERVAL, "1")`, and the second reason — `offsetContext.hasCompletelyProcessedPosition()` is always true once we have invoked `offsetContext.updateWalPosition`. So the number of heartbeats has nothing to do with a number of messages: once we have updated WAL position we will get heartbeats.

But when the fix is applied, the first heartbeat is produced right after `PgReplicationStream#readPending` is invoked (not after `INSERT INTO s1.a (aa) VALUES (11);`) (even though `PgReplicationStream#readPending` could return `null` not because keepalive message was received, but because there was no messages at all [1]). Because of this and because of `.with(Heartbeat.HEARTBEAT_INTERVAL, "1")` a lot of heartbeats are produced before change record is produced. This is why `expectedAtMostStartHeartbeats` is increased.

[1] We can change `PostgresReplicationConnection#readPending` method adding some additional check like:
```
final long lastReceiveLsnBeforeRead = stream.getLastReceiveLSN().asLong();
ByteBuffer read = stream.readPending();
final long lastReceiveLsn = stream.getLastReceiveLSN().asLong();

if (Objects.equals(lastReceiveLsn, lastReceiveLsnBeforeRead) {
  return false;
}
```
But I will point out that the specified test will still fail, because Postres will send keepalive message first when replication stream is started. Maybe this test should be rewritten to check if heartbeat starts when keepalive message is received?

Reply:
@jpechane, any thoughts on this? I'm a bit concerned about a flood of heartbeat events being produced. But I'm lacking insight into the details right now.

Reply:
@gunnarmorling Hi, the fact of increasing heartbeat messages is ok by itself. What the code does is increase the number opportunities to genereate heartbeat message, not number of heartbeat messages. And as the interval between heartebeat messages is set to 1 ms you see also increase the number of heartbeat messages too. If you set the interval to something more probable to use in production - probably minute and larger then the number of heartbeats will be very comparable.
I just wonder if the test itself should not be modified as we don't have a strict control over heartbeats now. So IMHO the correct aproach should be now
- write to the non-whitelisted table
- obtain LSN related to the change
- wait for 10 seconds or until a heartbeat message with offest with LSN >= LSN of change arrives
- verify that the slot has flush LSN set to vvalue >= LSN of change

Reply:
Yeah, sounds good to me. @igabaydulin, any chance you update the test along these lines? Thanks!

Reply:
@gunnarmorling No problem! I will update the test later as soon as I can :)